### DISCUSSION TITLE
Heads-Up: dropping CPython 2.7 in manylinux2010

### DISCUSSION THREAD

**Discussion by mayeut** [pypa/cibuildwheel#580] [General]
It's about time to drop support for CPython 2.7 !

`manylinux` will soon merge the `manylinux2014` branch in `master` which is used to build `manylinux2010`. There will now be a single branch / source of truth for what tools are installed in `manylinux` images with newer `PEP600` images on the way.

Rather than adding CPython 2.7 to `manylinux2014`, it will be dropped from `manylinux2010` once the merge happens.

If `cibuildwheel` wants to keep supporting CPython 2.7, we'll have to add specific pinned images for this use-case in `cibuildwheel`.

c.f. https://github.com/pypa/manylinux/issues/428#issuecomment-774541544

PS: `manylinux1` is not affected by this rewrite but support for `manylinux1` will probably be dropped by the summer.

**Comment by henryiii** [comment]
Nice site! (https://mayeut.github.io/manylinux-timeline/ ). Tiny note, your colors cycle, so the final four items on the Python versions have the same colors as the initial ones. MPL has a 9 or 10 color cycler by default, I think.

**Reply by mayeut** [reply]
> Tiny note, your colors cycle, so the final four items on the Python versions have the same colors as the initial ones. MPL has a 9 or 10 color cycler by default, I think.

I know, I've been too lazy to update this. I think I remember those were the ones automatically chose by `plotly` and I saw that but didn't update.

**Reply by henryiii** [reply]
That still looks like it's over 60% manylinux1, though? I've noticed at least NumPy is basing manylinux support by Python version. The 3.9 wheel comes in manylinux2010 only, while 3.7 and 3.8 still come in manylinux1. So manylinux2010 dropping 2.7 might just complicate things for a while, when it is still legal to build 2.7 + manylinux1. Once manylinux1 support drops, then cibuildwheel can drop 2.7 too, I'd think. Again, we work a lot harder to support 3.5 than 2.7, the "LTS" like-releases helped 2.7.

**Reply by mayeut** [reply]
Well, it's hard to know where exactly is the cursor for `manylinux1`. It's somewhere between 35% and 55%. Other than going over all projects manually, there's no way to tell if they built on `manylinux2010` or `manylinux2014` but still got `manylinux1`  compatibility (this is the case for my project which is really simple) or if they have multiple builds like is the case for NumPy.
I thought that NumPy made this choice based on the `pip` version available when 3.7, 3.8 & 3.9 got released but that does not seem consistent for 3.8, maybe @mattip has an idea ?
Although >99.9% manylinux systems downloading wheels are theoretically manylinux2010 compatible, `pip` might be too old to know about this. c.f. https://github.com/pypa/manylinux/issues/542#issuecomment-720139926

**Reply by henryiii** [reply]
I do believe it's almost purely pip and/or simply only providing manylinux1; in HEP we do have Scientific Linux 6 in use, but that should support manylinux2010 (just not 2014). But it comes with pip 7, IIRC. Pip 9 is common on CentOS 7, which is what we've replaced most of the 6's with. With 8 being killed off in favor of CentOS 8 Stream, I could see CentOS 7 sticking around till 2024... :'(

**Comment by henryiii** [comment]
When will 3.5 be dropped?

**Reply by henryiii** [reply]
Honestly, we have more workarounds for 3.5 than 2.7, since maintenance ended for that sooner than 2.7.

**Reply by mayeut** [reply]
I think 3.5 will drop after ubuntu 16.04 EOL, once python 3.10 beta lands. So probably around May, we'll drop CPython 3.5 and add CPython 3.10.

**Comment by mattip** [comment]
I want to drop manylinux1 for NumPy's next release (summer 2021). When NumPy does that, the other scientific python libraries will probably follow. 

> Pip 9 is common on CentOS 7, which is what we've replaced most of the 6's with

This is unfortunate. Is that true even for python 3.7+ ?

**Reply by mayeut** [reply]
> > Pip 9 is common on CentOS 7, which is what we've replaced most of the 6's with
>
> This is unfortunate. Is that true even for python 3.7+ ?

No, that's only true for the default pythons which are 2.7 & 3.6.
If you want something >= 3.7 on CentOS 7 and just want to yum install it, then there's python 3.8 from sclo which come with pip 19.3.1:
```
rh-python38-python-pip-19.3.1-1.el7.noarch : A tool for installing and managing Python packages
Repo        : centos-sclo-rh
Matched from:
Filename    : /opt/rh/rh-python38/root/usr/bin/pip3
```

**Reply by henryiii** [reply]
* 20.3 was the first version to support 3.9
* 19.1 or so was when 3.8 officially was supported
* 18.1 was the first version to support 3.7

Before that, it's not documented, but roughly 9 for 3.6, I expect, and I see a bug for 3.5 fixed in the 7 series. So by dropping support for 3.6, ~~you should not have any users stuck without the ability to use manylinux2010.~~

Edit: Oops, I though 2010 was added before 19. So there might be 18.x users. I don't think there are too many people stuck on 18, though - 9 was rather special, since 10 broke the `pip` script installed into sudo locations on many platforms. For 3.8+, this is true.

**Reply by henryiii** [reply]
Actually, a much better is to check the versions shipped by default: https://github.com/pypa/build/issues/226#issuecomment-775328927 

So 3.7.0 technically shipped with 10.

**Reply by mattip** [reply]
The problem is that manylinux1 is based on a very old CentOS. So while dropping manylinux1 may cause some pain, it is going to happen this year.

**Reply by henryiii** [reply]
I'm not worried[^1] about that, I just think cibuildwheel should match manylinux support. As soon as it's dropped, we should drop it too, basically.

[^1]: I'm not worried about that _here_. ;)

**Comment by mattip** [comment]
The problem with dropping manylinux1 is that pip will prefer to install-from-source than to use an older binary version by default. There is the `--only-binary :all:` flag, I wonder if we could teach CentOS7 to use that by default. Otherwise there will be many unhappy people who do `pip install` and end up compiling the world.

**Reply by Czaki** [reply]
> The problem with dropping manylinux1 is that pip will prefer to install-from-source than to use an older binary version by default. There is the `--only-binary :all:` flag, I wonder if we could teach CentOS7 to use that by default. Otherwise there will be many unhappy people who do `pip install` and end up compiling the world.

Could You explain or point documentation? I see no problem with installing a mix of manylinux1, manylinux2010, and manylinux2014 packages. Or you think about only `manylinux` images?

**Reply by mattip** [reply]
Say I have pip 9, that does not know about manylinux2010. When I do `pip install numpy`, it will see version 1.21 has a source tarball, but will not see the manylinux2010 binary wheels. Version 1.20 has a manylinux1 wheel, so theoretically pip could install that, but it will prefer the 1.21 source tarball over the 1.20 binary wheel. In order to convince this version of pip to install the 1.20 binary wheel, I need to do `pip install --only-binary numpy numpy`

**Reply by mayeut** [reply]
@Czaki,
to extend @mattip example, here's an excerpt from manylinux README, there's a minimal requirement on pip version to be able to install `manylinux*` wheels. `pip` will fail to "see" them on PyPI with versions less than the one listed here.
PEP600 is not yet in the readme and requires pip >= 20.3.

manylinux tag | Client-side pip version required
-- | --
manylinux2014 | pip >= 19.3
manylinux2010 | pip >= 19.0
manylinux1 | pip >= 8.1.0

**Reply by Czaki** [reply]
Ahh. "Upgrade pip first" problem...  
And mixing system python (outdated one) with working environment.

**Comment by mattip** [comment]
> No, that's only true for the default pythons which are 2.7 & 3.6.

Binary wheels for scientific python (NumPy, SciPy, ...) will not be shipped in 2021 for python<=3.7 anyway, so I think it is fine to deprecate manylinux1.

**Comment by joerick** [comment]
My feeling is that now we're over a year after EOL, we should consider dropping CPython 2.7 altogether, soon. We can still point in docs to our final release that supports it, and users that _really_ need it could run twice, like:

    CIBW_BUILD=cp27* pipx run cibuildwheel==1.9.0
    pipx run cibuildwheel==2.0.0

I suppose, in theory, we could accept bugfix PRs to a 1.9.x branch, if there's still a strong need when/if that breaks too.

Thoughts?

**Reply by Czaki** [reply]
what with pypy2.7

**Reply by mattip** [reply]
Why is it any different from CPython 2.7? This project's goal is to build wheels, I don't expect library authors to be building wheels for any python 2.7 anymore.

**Reply by Czaki** [reply]
pypy is still actively maintained?

**Reply by henryiii** [reply]
I don't think there's much cost in supporting 2.7? So therefore, we could support PyPy 2.7 but not CPython 2.7 without much extra cost if we want to. It's 3.5 that we pay a non-zero cost supporting, since, due to our design, we really care about things like ssl breaking, but Python 2 syntax doesn't bother us, we already only use the host Python.

Like @mattip said, though, I really don't expect libraries to build PyPy 2 wheels because they are not supporting Python 2 / CPython 2.7 anymore. But it would keep us from having to backport 7.3.4/7.4.0/8/whatever PyPy updates to the 1.9.x branch.

I'd really like to keep matching manylinux for support, personally. That means our manylinux2010 will drop 2.7 support when manylinux2010 does (duh), and we drop manylinux1 when manylinux1 dies this summer, and (C at least)Python 2.7 along with it. I think that would be the perfect point to drop 2.7 across Windows/macOS too. I really don't think we have any Windows/macOS 3.5 users, and likely not many Windows 2.7 users (where "user" is defined as people downloading the wheels, not libraries making the wheels!), so manylinux is the lowest common denominator for support.

Personally, I'd also be okay to backport critical bugfixes to 1.9/1.10/whatever.

**Reply by joerick** [reply]
> That means our manylinux2010 will drop 2.7 support when manylinux2010 does

Looks like that's today! Yesterday actually :P https://github.com/pypa/manylinux/issues/428#issuecomment-774729597

The cost for continued support is mostly existing complexity, special cases in the tests, etc. It doesn't really materialise until you have to bodge the tests to work around it e.g. the `exclude27` flag we added to `expected_wheels` recently. Not exactly a huge burden either though. 

I haven't worked in Python 2.7 much lately anyway, but if we were following the data, according to @mayeut's site, 20% of projects are still publishing py2 wheels, and Python2 [still makes up 10% of the downloads](https://pypistats.org/packages/__all__). I bet most of those 20% of projects are only shipping them because 'it still works, why not ¯\\\_(ツ)\_/¯', but I am curious about that 10% of PyPI downloads. Any reason (other than fully supporting manylinux1 until its retirement) that you'd like to keep 2.7 around @henryiii ? Not that I disagree necessarily, just curious...

**Reply by henryiii** [reply]
Primarily, I don't want users to have a reason to roll their own system instead of using cibuildwheel; not supporting something that you otherwise can get if you do it yourself is a reason not to use cibuildwheel. Having a "LTS" 1.9.x series (or 1.10.x if there's something else to get in - the Limited ABI stuff is inherently Python 3, so that doesn't matter) could be a fine path forward, though. And if we don't follow manylinux, then we have to invent our own support dates - I'd rather just leave it to manylinux since they seem to be pretty reasonable, and we can't easily expected do more than they do. (and most "old" users tend to be manylinux, along with maybe a few default "python" on macOS users).

Personally, since you asked, there are a few things dragging me back on Python 2. One is I have been promising a final 2.7 release of boost-histogram since late last year that's API compatible with the 3+ only release coming out soon. The other is that awkward is still 2.7 compatible (some portion of it, maybe uproot only, is actually secretly 2.6 compatible), and I want to help move it over to cibuildwheel. I think, by mid-year, all new Scikit-HEP package releases (with binaries, anyway) will likely be 3+ only. I'm assuming some other projects are in a similar boat, especially in the sciences, where users are not that keen to use the latest features, and it's a hassle to upgrade. It has to happen, but a manylinux drop date is one that's really hard to argue with.

Arch emulation and Universal2 where the key things to get in, though, so moving forward is also an option I would be okay with - I can avoid upgrading off of 1.9.x for a little while in a few projects if need be.

**Reply by YannickJadoul** [reply]
I was in a similar situation as @henryiii, until this weekend: I had promised a last 2.7 version, but kept dragging my feet on actually releasing it. Feels good to finally have that, though ;-) (Ideally, if I still have bug fixes, I'd still release those patch versions in 2.7 as well, btw; but luckily I don't write bugs, of course ;-) )

But I think the argument about not forcing users to roll their own is a very good one to consider! That, combined with the fact that `cibuildwheel` is not supposed to be making that decision (but rather follow the core `pip`, `wheel`, `manylinux` projects, where people have probably had much longer discussions about it?)

At the same time, it's also good to realize that if we keep supporting it, that might be contributing to the prolonging of pain, though? I do expect that lots of project (to be fair, mine as well) would just provide 2.7 wheels because it's so easy?

**Reply by henryiii** [reply]
Pip has dropped 2.7, and wheel is likely going to have to soon (as packaging will be dropping it; unless they pin packaging). But manylinux is the key thing we are sitting on top of, so that's the one I think should follow. If we drop 2.7, there's a chance packages would just go back to rolling their own if they really felt 2.7 was important. But if manylinux drops it (and/or wheel), then it's not "our" fault, and they are less likely to want to drop us. :)

PS: Most packages release manylinux1 wheels; NumPy _just_ stopped releasing manylinux1 for NumPy on 3.9; the others (Matplotlib and such) are still manylinux1.

**Reply by Czaki** [reply]
Some packages still release fake manylinux1 packages (like PySide2)

**Reply by henryiii** [reply]
I do too. ;)

**Reply by henryiii** [reply]
Though, in my case, it's only "fake" due to compiling a new compiler for it, as GCC 4.8 didn't support C++14.

**Reply by Czaki** [reply]
Qt in the new versions needs a newer version of glibc than could be assumed with manylinux1 tag.

**Reply by YannickJadoul** [reply]
> Though, in my case, it's only "fake" due to compiling a new compiler for it, as GCC 4.8 didn't support C++14.

I _cough cough cough_ used to just statically link libstdc++, actually, to solve that issue (see https://github.com/pypa/manylinux/issues/118). Then decided it's not worth it and `manylinux2010` was a lot more relaxing.
